asm

tcheremkhina opened the pull request.

tcheremkhina committed:
asm hw, first try

tcheremkhina committed:
no important changes

kt-cpp commented on asm/sub.asm:
```diff
@@ -0,0 +1,392 @@
+                mov             rax, [rsi]
+                lea             rsi, [rsi + 8]
+                add             rax, r10
+                sub             [rdi], rax
+
+                jnc             .ok
+                mov             r10, 1
+                jmp             .r
```
Обычно такое делается через ```sbb```, но так как здесь решение, которое выводит вычитание по модулю и такая самописная штука, то пусть будет так

kt-cpp commented on asm/mul.asm:
```diff
@@ -0,0 +1,391 @@
+; result:
+;   [rdi] = [rdi] * [rsi]
+mul_long_long:
+                push            rdi
```
Зачем пушить ```rdi``` если все равно его портишь

tcheremkhina replied:
fixed

kt-cpp commented on asm/mul.asm:
```diff
@@ -0,0 +1,391 @@
+                pop             r14
+                ret
+
+; multiplies two long numbers
+;   rdi -- address of num1 (long number)
+;   rsi -- address of num2 (long number)
+;   rcx -- length of numbers in qwords
+; result:
+;   [rdi] = [rdi] * [rsi]
```
Неявным аргументом к этой функции является ```r11```, где ты хранишь зарезервированное пространство для вычислений

И ответ на самом деле тоже зарезервирован

И почему нельзя портить стэк в этой функции для ```tmp``` переменной? Кажется это просто

tcheremkhina replied:
Непонятно что требуется в этом комментарии. 
В третьем замечании непонятно о чем речь.

kt-cpp replied:
Про третье: почитай про стековый фрейм и как выделяются локальные переменные на стеке. Подумай о том, можно ли сделать твой tmp буффер локальной переменной.

Первые два замечания видимо понятны

Требуется: написать так, чтобы когда я посмотрел на контракт функции - я понял, что и в какие регистры мне положить. И мне не хочется класть еще что-то дополнительное - связанное с твоими вычислениями внутри функции (tmp буффер).

tcheremkhina replied:
вроде fixed

kt-cpp commented on asm/mul.asm:
```diff
@@ -0,0 +1,391 @@
+
+                sub             rsp, 6 * 256 * 8
+                lea             rdi, [rsp + 256 * 8]
+                mov             rcx, 256
```
Тесты расчитаны на 128 разрядов * 128 разрядов, поэтому они не видят ошибку, а я вижу

Умножение пока что не рабочее (фиксится удалением одной пары строчек, думаю найдете :)

tcheremkhina replied:
fixed

kt-cpp commented on asm/mul.asm:
```diff
@@ -0,0 +1,391 @@
+; copies long to long
+;   r14 -- address of num1 (long number)
+;   rdi -- address of num2 (long number)
+;   r12 -- length of numbers in qwords
```
А давайте ```rcx``` для консистентности

tcheremkhina replied:
fixed

kt-cpp reviewed (changes requested):
Есть много мелочей, но пока что они опущены, потому что умножение не рабочее

tcheremkhina committed:
multiply fixed

tcheremkhina committed:
fixed

kt-cpp commented on asm/mul.asm:
```diff
@@ -0,0 +1,390 @@
+                push            rsi
+                push            r11
+
+                mov             rcx, 256
```
`rcx` портится после выхода из этой функции

kt-cpp replied:
И он никак в этой строчке не зависит от входного `rcx`

tcheremkhina replied:
Теперь зависит от входного

kt-cpp commented on asm/mul.asm:
```diff
@@ -0,0 +1,390 @@
+
+                mov             rcx, 256
+
+                sub             rsp, 2 * 256 * 8
```
А не многовато ли доппамяти?

tcheremkhina replied:
Исправила

kt-cpp reviewed (changes requested):
Памяти использовано `5 * 256 * 8` байт, чтобы считать два 128 разрядных числа, перемножить их и вывести 256р ответ

Можно сократить почти в 2 раза (`mul_long_short` даст тебе максимум на один разряд больше, и хранить входные числа как 256р слишком overhead)

tcheremkhina committed:
memory optimized

tcheremkhina committed:
tests for sub added
sub fixed

tcheremkhina committed:
sub fixed

Извините за коммиты после отправки заявки, просто внезапно оказалось что тесты на sub не были добавлены и sub нерабочий.

lejabque commented:
Выглядит так, как будто после ребейза из upstream ты сделала `git push -f origin asm` раньше, чем `git push -f origin master`.
Попробуй сделать так:
```
git checkout asm
git rebase -i HEAD~20
// откроется текстовый редактор стандартный, в нём нужно у всех коммитов, 
// которые должны быть в мастере (сделаны не тобой) поменять pick на d, затем сохранить этот файл и выйти из него
git push -f origin asm
git rebase master
git push -f origin asm
```

После этого, по идее, должно всё стать норм, если не получится, то телеграм @lejabq

kt-cpp reviewed (approved):
Ок.
